### ISSUE TITLE
Implementation Plan: Pure Isabelle/HOL Neural Network Library

### ISSUE THREAD

**Issue by danregima** [ReZorg/nn.nn#12]
<div class="markdown-body MarkdownRenderer-module__container__sJ8jf" data-copilot-markdown="true" style="--MarkdownRenderer_streaming-fade-duration: 750ms;"><h1>Implementation Plan: Pure Isabelle/HOL Neural Network Library (<code node="[object Object]">lang/isa</code>)</h1>
<h2>Goal</h2>
<p>Add a pure Isabelle/HOL implementation of the neural network library as a new language subtree <code node="[object Object]">lang/isa/</code>, following the established repository pattern (core library + tests + demo + example + README + root summary + CI job). The theory files are <strong>derived from two existing sources</strong>:</p>
<ol>
<li node="[object Object]"><strong>The formal Z++ specifications</strong> in <code node="[object Object]">docs/</code> (<code node="[object Object]">data_model.zpp</code>, <code node="[object Object]">system_state.zpp</code>, <code node="[object Object]">operations.zpp</code>, <code node="[object Object]">integrations.zpp</code>) — these already define types, invariants, and operational semantics in a form that maps almost one-to-one onto Isabelle definitions and lemmas. This finally realizes the promise in <code node="[object Object]">FORMAL_SPEC_INDEX.md</code> ("Use Z++ specs for theorem proving").</li>
<li node="[object Object]"><strong>The sibling functional implementations</strong> (<code node="[object Object]">lang/scm/nn.scm</code>, <code node="[object Object]">lang/rkt/nn.rkt</code>, <code node="[object Object]">lang/raku/NN.rakumod</code>) — these fix the concrete API surface (~70 functions), data representations (lists of lists), and the 45-test suite to mirror.</li>
</ol>
<p>Isabelle is unique among the ports: besides an executable implementation, it can <strong>prove</strong> the invariants that the Z++ specs only state.</p>
<h2>Key Design Decisions</h2>
<ol>
<li node="[object Object]">
<p><strong>Directory and session layout</strong>: <code node="[object Object]">lang/isa/</code> containing a <code node="[object Object]">ROOT</code> file defining one Isabelle session <code node="[object Object]">NN</code> (based on <code node="[object Object]">HOL</code>), with theory files listed below. Build/test entry point is <code node="[object Object]">isabelle build -D lang/isa</code>.</p>
</li>
<li node="[object Object]">
<p><strong>Pure Isabelle/HOL, no AFP</strong>: depend only on the Isabelle distribution (<code node="[object Object]">Main</code>/<code node="[object Object]">Complex_Main</code> and <code node="[object Object]">HOL-Library</code>). No Archive of Formal Proofs entries, keeping the implementation "pure" and CI-installable.</p>
</li>
<li node="[object Object]">
<p><strong>Data representation mirrors the siblings</strong>: vectors as <code node="[object Object]">real list</code>, matrices as <code node="[object Object]">real list list</code>, neurons as weight-list + bias pairs, layers as neuron lists, networks as layer-size list + weights — matching the Scheme/Racket/Raku structures and <code node="[object Object]">data_model.zpp</code> Sections 1–4 rather than HOL's fixed-dimension <code node="[object Object]">vec</code> types. This keeps everything executable and structurally parallel to the other ports.</p>
</li>
<li node="[object Object]">
<p><strong>Executability strategy (the main Isabelle-specific problem)</strong>: <code node="[object Object]">exp</code> on <code node="[object Object]">real</code> is not executable. Use a two-level approach:</p>
<ul>
<li node="[object Object]">Prove all properties against the exact <code node="[object Object]">real</code>-valued definitions.</li>
<li node="[object Object]">Make demos/tests runnable via <code node="[object Object]">value</code>/<code node="[object Object]">eval</code> by importing <code node="[object Object]">HOL-Library.Code_Real_Approx_By_Float</code> in the test/demo theories only (keeping proof theories float-free), or alternatively by providing a truncated-Taylor-series rational <code node="[object Object]">exp-safe</code> (the siblings already clamp <code node="[object Object]">exp</code> to ±20, so a bounded approximation is faithful). Decide on the float-approximation route as primary; document the trade-off in the README.</li>
</ul>
</li>
<li node="[object Object]">
<p><strong>Determinism instead of randomness</strong>: Isabelle is pure, so replace <code node="[object Object]">random-real</code>/<code node="[object Object]">random-weight</code> with a seeded linear congruential generator threaded through weight initialization. This mirrors the siblings' init-in-(−0.5, 0.5) contract from <code node="[object Object]">integrations.zpp</code> Section 2 while making all demos reproducible.</p>
</li>
<li node="[object Object]">
<p><strong>Module system as a deep embedding</strong>: define the module zoo (<code node="[object Object]">Linear</code>, <code node="[object Object]">Identity</code>, <code node="[object Object]">Sigmoid</code>, <code node="[object Object]">Tanh</code>, <code node="[object Object]">ReLU</code>, <code node="[object Object]">Softmax</code>, <code node="[object Object]">LogSoftmax</code>, <code node="[object Object]">Reshape</code>, <code node="[object Object]">Mean</code>, <code node="[object Object]">Max</code>, <code node="[object Object]">Sequential</code>, <code node="[object Object]">Concat</code>) as one recursive datatype with a primitive-recursive forward function, and criterions (<code node="[object Object]">MSE</code>, <code node="[object Object]">ClassNLL</code>, <code node="[object Object]">BCE</code>, <code node="[object Object]">Abs</code>) as a second datatype — the direct Isabelle rendering of <code node="[object Object]">data_model.zpp</code> Sections 5–8 and 10.</p>
</li>
</ol>
<h2>Theory Files (with derivation traceability)</h2>
<table><thead><tr><th>Theory</th><th>Derived from</th><th>Contents</th></tr></thead><tbody><tr><td><code node="[object Object]">NN_Types.thy</code></td><td><code node="[object Object]">data_model.zpp</code> §1–3</td><td>Type synonyms for vector/matrix/shape; neuron, layer, sample records; well-formedness predicates (shape consistency invariants)</td></tr><tr><td><code node="[object Object]">NN_Vector.thy</code></td><td><code node="[object Object]">integrations.zpp</code> §1; sibling vector ops</td><td>dot-product, add, sub, scalar-mult, sum, mean, max, map/map2</td></tr><tr><td><code node="[object Object]">NN_Matrix.thy</code></td><td><code node="[object Object]">integrations.zpp</code> §1</td><td>matrix-vector multiplication, transpose, dimension predicates</td></tr><tr><td><code node="[object Object]">NN_Activations.thy</code></td><td><code node="[object Object]">data_model.zpp</code> §7; <code node="[object Object]">integrations.zpp</code> §2</td><td>exp-safe (clamped), sigmoid, tanh, relu + their derivatives; softmax and log-softmax (numerically stable, max-subtracted)</td></tr><tr><td><code node="[object Object]">NN_Network.thy</code></td><td><code node="[object Object]">data_model.zpp</code> §4; <code node="[object Object]">operations.zpp</code> §1, §6</td><td>seeded weight init, create-network, neuron/layer output, forward-propagate, forward, predict</td></tr><tr><td><code node="[object Object]">NN_Loss.thy</code></td><td><code node="[object Object]">data_model.zpp</code> §8; <code node="[object Object]">operations.zpp</code> §2</td><td>MSE + derivative, ClassNLL, BCE, Abs losses</td></tr><tr><td><code node="[object Object]">NN_Backprop.thy</code></td><td><code node="[object Object]">operations.zpp</code> §3–4</td><td>layer activations cache, output-layer deltas, gradient computation, SGD weight/bias update</td></tr><tr><td><code node="[object Object]">NN_Training.thy</code></td><td><code node="[object Object]">operations.zpp</code> §5; <code node="[object Object]">system_state.zpp</code> §3</td><td>train-step, train-epoch, train (epoch-count recursion, learning-rate parameter)</td></tr><tr><td><code node="[object Object]">NN_Modules.thy</code></td><td><code node="[object Object]">data_model.zpp</code> §5–6, §10</td><td>module datatype, module-forward, sequential/concat semantics, criterion datatype, criterion-forward</td></tr><tr><td><code node="[object Object]">NN_Properties.thy</code></td><td>invariants throughout all four <code node="[object Object]">.zpp</code> files</td><td>The proof payload (see below)</td></tr><tr><td><code node="[object Object]">NN.thy</code></td><td>—</td><td>Umbrella theory importing everything (the "library")</td></tr><tr><td><code node="[object Object]">NN_Test.thy</code></td><td>sibling 45-test suites</td><td>Executable regression tests as <code node="[object Object]">eval</code>-proved lemmas</td></tr><tr><td><code node="[object Object]">NN_Demo.thy</code> / <code node="[object Object]">NN_Examples.thy</code></td><td>sibling <code node="[object Object]">demo.*</code>/<code node="[object Object]">example.*</code></td><td>XOR and AND-gate training runs, module composition, softmax classification demos via <code node="[object Object]">value</code></td></tr></tbody></table>
<h2>Proof Obligations (<code node="[object Object]">NN_Properties.thy</code>) — the derived invariants</h2>
<p>Prove, as Isabelle lemmas, the invariants the Z++ specs assert, each annotated with its spec section:</p>
<ul>
<li node="[object Object]"><strong>Activation ranges</strong> (<code node="[object Object]">data_model.zpp</code> §7): sigmoid output strictly in (0,1); tanh in (−1,1); relu non-negative; relu is identity on non-negatives.</li>
<li node="[object Object]"><strong>Softmax correctness</strong> (<code node="[object Object]">data_model.zpp</code> §7 invariant): components non-negative and sum to 1 for non-empty input; log-softmax = log ∘ softmax.</li>
<li node="[object Object]"><strong>Shape preservation</strong> (<code node="[object Object]">data_model.zpp</code> §2, §11 invariants; <code node="[object Object]">operations.zpp</code> §1): layer output length equals neuron count; forward through a well-formed network yields output of the declared final layer size; sequential module forward preserves declared shapes.</li>
<li node="[object Object]"><strong>Algebraic properties</strong> (<code node="[object Object]">integrations.zpp</code> §1): dot-product commutativity/linearity; transpose involution; matrix-vector dimension contracts.</li>
<li node="[object Object]"><strong>Loss properties</strong> (<code node="[object Object]">operations.zpp</code> §2): MSE non-negative, zero iff prediction equals target (on equal-length vectors); Abs loss non-negative.</li>
<li node="[object Object]"><strong>Derivative correctness</strong> (<code node="[object Object]">integrations.zpp</code> §2): sigmoid-derivative equals the analytic derivative of sigmoid (via <code node="[object Object]">DERIV</code>/<code node="[object Object]">has_real_derivative</code>); same for tanh; MSE-derivative correct.</li>
<li node="[object Object]"><strong>Training preserves structure</strong> (<code node="[object Object]">system_state.zpp</code> §5 global invariants): train-step/train-epoch/train preserve network topology (layer sizes unchanged) and well-formedness.</li>
</ul>
<h2>Deliverables Checklist (implementation order)</h2>
<ol>
<li node="[object Object]"><strong>Scaffolding</strong>: <code node="[object Object]">lang/isa/ROOT</code> (session <code node="[object Object]">NN</code>, document off), directory layout.</li>
<li node="[object Object]"><strong>Core theories</strong>: <code node="[object Object]">NN_Types</code>, <code node="[object Object]">NN_Vector</code>, <code node="[object Object]">NN_Matrix</code>, <code node="[object Object]">NN_Activations</code> — verify with <code node="[object Object]">isabelle build</code> early.</li>
<li node="[object Object]"><strong>Network + forward pass</strong>: <code node="[object Object]">NN_Network</code> (including deterministic PRNG init).</li>
<li node="[object Object]"><strong>Loss, backprop, training</strong>: <code node="[object Object]">NN_Loss</code>, <code node="[object Object]">NN_Backprop</code>, <code node="[object Object]">NN_Training</code>.</li>
<li node="[object Object]"><strong>Module system</strong>: <code node="[object Object]">NN_Modules</code> with the two datatypes and forward semantics.</li>
<li node="[object Object]"><strong>Proofs</strong>: <code node="[object Object]">NN_Properties</code> covering the table above; keep lemma names mirroring zpp invariant names for traceability.</li>
<li node="[object Object]"><strong>Tests</strong>: <code node="[object Object]">NN_Test</code> mirroring the sibling 45-test categories (activations, vector ops, reductions, network creation, forward, losses, each module type, criterions, softmax, integration/training convergence smoke checks) as machine-checked <code node="[object Object]">eval</code> lemmas.</li>
<li node="[object Object]"><strong>Demos/examples</strong>: <code node="[object Object]">NN_Demo</code> (XOR, AND gate, module composition, activation/loss comparisons) and <code node="[object Object]">NN_Examples</code>, producing output via <code node="[object Object]">value</code> so a session build exercises them end-to-end. Optionally demonstrate <code node="[object Object]">export_code</code> to SML/Haskell as a bonus "derived implementation" artifact.</li>
<li node="[object Object]"><strong>Documentation</strong>:
<ul>
<li node="[object Object]"><code node="[object Object]">lang/isa/README.md</code> following the sibling README structure (features, install/run with <code node="[object Object]">isabelle build</code>, quick start, API reference, examples, architecture, torch/nn comparison table, spec-to-theory traceability matrix).</li>
<li node="[object Object]">Root <code node="[object Object]">IMPLEMENTATION_ISABELLE.md</code> following <code node="[object Object]">IMPLEMENTATION_SCHEME.md</code>/<code node="[object Object]">IMPLEMENTATION_RACKET.md</code> format.</li>
<li node="[object Object]">Update <code node="[object Object]">README.md</code> CI table with the Isabelle row (<code node="[object Object]">cd lang/isa &amp;&amp; isabelle build -D .</code>).</li>
<li node="[object Object]">Update <code node="[object Object]">FORMAL_SPEC_INDEX.md</code> to note the specs are now mechanized in <code node="[object Object]">lang/isa</code>.</li>
</ul>
</li>
<li node="[object Object]"><strong>CI integration</strong> (<code node="[object Object]">.github/workflows/ci.yml</code>):</li>
</ol>
<ul>
<li node="[object Object]">Add <code node="[object Object]">isa</code> entry to the <code node="[object Object]">changes</code> paths-filter (<code node="[object Object]">lang/isa/**</code>).</li>
<li node="[object Object]">New <code node="[object Object]">isabelle</code> job: install Isabelle (download a pinned release tarball, cached with <code node="[object Object]">actions/cache</code> including heap images since the ~1 GB download and HOL heap build dominate runtime), then <code node="[object Object]">isabelle build -o document=false -v -D lang/isa</code>. A successful build simultaneously type-checks the library, re-runs every proof, and executes all <code node="[object Object]">eval</code>-based tests and <code node="[object Object]">value</code> demos — build success is the test gate.</li>
<li node="[object Object]">Add the job to the <code node="[object Object]">ci-success</code> fan-in <code node="[object Object]">needs</code> list.</li>
</ul>
<h2>Risks / Edge Cases to Handle</h2>
<ul>
<li node="[object Object]"><strong><code node="[object Object]">exp</code>/<code node="[object Object]">tanh</code> executability</strong>: confine float code-generation setup to test/demo theories so proof theories stay sound; clamp behavior must match the siblings' ±20 cutoff.</li>
<li node="[object Object]"><strong>Division/log partiality</strong> (BCE, ClassNLL, log-softmax): HOL total functions make these safe by default, but lemmas must carry the positivity side-conditions the Z++ specs state (<code node="[object Object]">integrations.zpp</code> §6 numerical stability).</li>
<li node="[object Object]"><strong>Empty vectors / ragged matrices</strong>: guard lemmas with well-formedness predicates from <code node="[object Object]">NN_Types</code>; mirror sibling behavior for degenerate inputs.</li>
<li node="[object Object]"><strong>Training convergence tests</strong>: unlike sibling suites, don't assert exact learned values (float nondeterminism across Isabelle versions); assert structural properties and loss-decrease on the fixed-seed XOR run instead.</li>
<li node="[object Object]"><strong>CI runtime</strong>: pin one Isabelle version (e.g., Isabelle2025) rather than a matrix; cache <code node="[object Object]">~/.isabelle</code> heaps keyed on the version + ROOT hash.</li>
</ul></div>
